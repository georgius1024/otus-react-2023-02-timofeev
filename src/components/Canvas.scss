.canvas {
  position: relative;
  outline: 4px solid #ccc;
  outline-offset: 2px;
  background-color: #333;
  color: #ccc;
  overflow: hidden;
  .cell {
    position: absolute;
    background-color: #ccc;
    border-radius: 100%;
    transition: 200ms ease;
    &:hover {
      background-color: #aaa;
      transform: scale(2);
    }
  }
}