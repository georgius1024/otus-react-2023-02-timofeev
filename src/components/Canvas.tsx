import "@/components/Canvas.scss";
import { ReactElement } from "react";
import Dot from "@/components/Dot";

type clickHandler = (col: number, row: number) => void;
type CanvasProps = {
  cols: number;
  rows: number;
  dx: number;
  dy: number;
  onClick: clickHandler;
};

export default function Canvas(props: CanvasProps): ReactElement {
  const canvasStyle = {
    width: `${props.cols * props.dx}px`,
    height: `${props.rows * props.dy}px`,
  };

  const cols = (row: number) => {
    return new Array(props.cols).fill(0).map((_, col) => {
      return (
        <Dot
          col={col}
          row={row}
          key={`${row}-${col}`}
          cols={props.cols}
          rows={props.rows}
          dx={props.dx}
          dy={props.dy}
          onClick={() => props.onClick(col, row)}
        />
      );
    });
  };
  const rows = () => {
    return new Array(props.rows)
      .fill(0)
      .map((_, row) => cols(row))
      .flat();
  };

  return (
    <div className="canvas" style={canvasStyle}>
      {rows()}
    </div>
  );
}
