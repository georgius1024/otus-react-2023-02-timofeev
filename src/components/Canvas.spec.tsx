import { render, screen } from "@testing-library/react";
import { vi, expect } from "vitest";
import Canvas from "@/components/Canvas";

describe("Canvas", () => {
  const clickHandler = vi.fn();
  let wrapper: any;
  beforeEach(() => {
    wrapper = render(<Canvas cols={3} rows={3} dx={50} dy={50} onClick={clickHandler} />);
  });
  afterEach(() => {
    clickHandler.mockReset();
  });
  it("renders without errors", () => {
    expect(wrapper).toMatchSnapshot();
  });
  it("dot clicks emits events", () => {
    const dots = screen.queryAllByTestId("cell-dot-marker");
    expect(dots).toHaveLength(9);
    expect(dots[0]).toBeVisible();
    dots[0].click();
    expect(clickHandler).toHaveBeenCalled();
  });
});
