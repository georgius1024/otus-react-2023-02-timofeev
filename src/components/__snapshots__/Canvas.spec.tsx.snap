// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Canvas > renders without errors 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="canvas"
        style="width: 150px; height: 150px;"
      >
        <div
          class="cell"
          data-testid="cell-dot-marker"
          style="left: 1px; top: 1px; width: 48px; height: 48px;"
        />
        <div
          class="cell"
          data-testid="cell-dot-marker"
          style="left: 51px; top: 1px; width: 48px; height: 48px;"
        />
        <div
          class="cell"
          data-testid="cell-dot-marker"
          style="left: 101px; top: 1px; width: 48px; height: 48px;"
        />
        <div
          class="cell"
          data-testid="cell-dot-marker"
          style="left: 1px; top: 51px; width: 48px; height: 48px;"
        />
        <div
          class="cell"
          data-testid="cell-dot-marker"
          style="left: 51px; top: 51px; width: 48px; height: 48px;"
        />
        <div
          class="cell"
          data-testid="cell-dot-marker"
          style="left: 101px; top: 51px; width: 48px; height: 48px;"
        />
        <div
          class="cell"
          data-testid="cell-dot-marker"
          style="left: 1px; top: 101px; width: 48px; height: 48px;"
        />
        <div
          class="cell"
          data-testid="cell-dot-marker"
          style="left: 51px; top: 101px; width: 48px; height: 48px;"
        />
        <div
          class="cell"
          data-testid="cell-dot-marker"
          style="left: 101px; top: 101px; width: 48px; height: 48px;"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="canvas"
      style="width: 150px; height: 150px;"
    >
      <div
        class="cell"
        data-testid="cell-dot-marker"
        style="left: 1px; top: 1px; width: 48px; height: 48px;"
      />
      <div
        class="cell"
        data-testid="cell-dot-marker"
        style="left: 51px; top: 1px; width: 48px; height: 48px;"
      />
      <div
        class="cell"
        data-testid="cell-dot-marker"
        style="left: 101px; top: 1px; width: 48px; height: 48px;"
      />
      <div
        class="cell"
        data-testid="cell-dot-marker"
        style="left: 1px; top: 51px; width: 48px; height: 48px;"
      />
      <div
        class="cell"
        data-testid="cell-dot-marker"
        style="left: 51px; top: 51px; width: 48px; height: 48px;"
      />
      <div
        class="cell"
        data-testid="cell-dot-marker"
        style="left: 101px; top: 51px; width: 48px; height: 48px;"
      />
      <div
        class="cell"
        data-testid="cell-dot-marker"
        style="left: 1px; top: 101px; width: 48px; height: 48px;"
      />
      <div
        class="cell"
        data-testid="cell-dot-marker"
        style="left: 51px; top: 101px; width: 48px; height: 48px;"
      />
      <div
        class="cell"
        data-testid="cell-dot-marker"
        style="left: 101px; top: 101px; width: 48px; height: 48px;"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
